### ISSUE TITLE
[FEATURE] Implement content filters

### ISSUE THREAD

**Issue by mihail-varbanov** [rokwire/illinois-app#2926]
**Re: University Monday 2023/01/16** / 2022-01-17

2 Misho, we want to implement "filters" that will be used in Groups initially. We are thinking the JSON would be stored in Content BB. Initially we need to specify College and Department when a Group is created/modified. Here is what Stephen and I came up with:

```
Create Group UI:
Group Name:
Description:
Filters:
<START DYNAMIC CONTENT FROM CONTENT BB>
College<label>: ___Grainger____<dropdown> Department<label>: ____ECE___<dropdown> Major<label>: _______<dropdown>
Class<label>: _______<dropdown>
<END>

```
```
Group Model (API/DB):
{
"name": "ECE Freshmen",
...
"filters": {
"college": "Grainger",
"college.department": "ECE",
"class": "2027"
}
}

```
```
Content BB JSON:
[
{
"id": "college",
"label": "College",
"multi-select": false,
"default_values": ["Grainger", "Gies"]
},
{
"id": "college.department",
"label": "Department",
"multi-select": false,
"parent_label": "College",
"default_values": null,
"subvalues": {
"Grainger": ["ECE", "CS", "MechE"],
"Gies": ["Accounting", "Finance"]
}
},
{
"id": "college.department.major",
"label": "Major",
"multi-select": false,
"parent_label": "Department",
"default_values": null,
"subvalues": {
"ECE": ["CompE", "EE"]
}
},
{
"id": "class",
"label": "Class",
"multi-select": true,
"default_values": ["2023", "2024"]
}
]
```

Linked to rokwire/app-flutter-plugin#246

**Comment by mihail-varbanov** [comment]
**Re: University Monday 2023/01/16** / 2022-01-17

MV: I would suggest the following structure in Content BB but can't this wait until Maps2 gets finished?

```
[{
  "id": "college",
  "label": "College",
  "multi-select": false,
  "values": [
    { "id": "college.grainger", "label": "Grainger" },
    { "id": "college.gies",     "label": "Gies" }
  ]
},{
  "id": "college.department",
  "label": "Department",
  "multi-select": false,
  "values": [
    { "id": "college.department.ece", "label": "ECE", "requres": { "college" : "college.grainger" }},
    { "id": "college.department.cs", "label": "CS", "requres": { "college" : "college.grainger" }},
    { "id": "college.department.mech_e", "label": "MechE", "requres": { "college" : "college.grainger" }}
    { "id": "college.department.accounting", "label": "Accounting", "requres": { "college" : "college.gies" }},
    { "id": "college.department.finance", "label": "Finance", "requres": { "college" : "college.gies" }}
  ]
},{
  "id":"college.department.major",
  "label":"Major",
  "multi-select":false,
  "values":{
    { "id":"college.department.major.comp_e", "label": "CompE", "requres": { "college.department" : "college.department.ece" }},
    { "id":"college.department.major.ee", "label": "EE", "requres": { "college.department" : "college.department.ece" }}
  }
},{
  "id":"class",
  "label":"class",
  "multi-select":true,
  "values":[
    { "id":"class.2023", "label": "2023" },
    { "id":"class.2024", "label": "2024" }
  ]
}]

```

**Comment by mihail-varbanov** [comment]
**Re: University Monday 2023/01/16** / 2022-01-18

MV: In addition, instead of having the "multi-select" boolean field we would better define "min-select-count" and "max-select-count" integer fields that give better control (e.g. we can require multiple selection up to 5 items, for example). We have worked this out in research questionnaire answers.

**Comment by mihail-varbanov** [comment]
**Re: University Monday 2023/01/16** / 2022-01-19

1.1 Stephen and I talked about your Filters JSON and your latest comment on using ...-select-count and we believe this is more flexible than our original. Thanks.

**Comment by mihail-varbanov** [comment]
**University Monday 2023/01/23** / 2022-01-23

I think we should have a button like Tags or Advanced Settings called Filters so the Group page is not so long, thoughts?

**Comment by mihail-varbanov** [comment]
**Re: University Monday 2023/01/23** / 2022-01-23

Something else that I would like to discuss with you and missed to mention in my email from Friday was the format of filters that we store in group objects.

You proposed this:
```
Group Model (API/DB):
{
  "name": "ECE Freshmen",
  ...
  "filters": {
    "college": "Grainger",
    "college.department": "ECE",
    "class": "2027"
  }
}
```
^ The key is a filter id, the value is option title.

I am thinking about this:
```
{
  "name": "ECE Freshmen",
  ...
  "filters": {
    "college": "college.grainger",
    "college.department": "college.department.ece",
    "class": "class.2027"
  }
}
```
^ The difference is how we store values that are value IDs here instead of value titles above (you can see the current filters definition [here](https://github.com/rokwire/illinois-app/blob/develop/assets/content.filters.groups.json)).

The client currently represents filter selection internally as a map of filter_id: option_id. However it stores it as a map filter_id : option_title as you initially proposed and makes a coverson between these two models each time it loads or stores a group. From a client perspective it is much easier to work with filter_id: option_id mapping when maintaining selection. Also, option titles can vary across different languages but let's assume that we will store them in English.

The question which format is more suitable depends on who other that the client will work with group filters and how they will be acknowledged anywhere else. The problem with filter_id  : option_id mapping is that we cannot display group filter values without loading the filter definitions first, which is not convenient when we display group details in a card widget or detail panel. However this representation is more convenient if we filter a group's display list from filters dropdown. Also please note that currently we actually use some mixed approach because we use ID to identify a filter and title to identify an option.

So I am wondering which selection format would apply best for what we need. Can you please share your thoughts or shed some light how these filters will be used elsewhere than in the create/edit group panel?

**Comment by mihail-varbanov** [comment]
**Re: University Monday 2023/01/23** / 2022-01-24

4.1 Go ahead with your proposals on using IDs etc.

**Comment by mihail-varbanov** [comment]
**Re: University Monday 2023/01/23** / 2022-01-24

Additionally I wanted to mention that we will need on the All/My Groups panel we will want to be able to select and apply filters to these views. Show All/My Groups for the College of Engineering (shows all departments) and College of Engineering/Department of Robotics etc

**Comment by mihail-varbanov** [comment]
**University Monday 2023/01/30** / 2023-01-30

1.1 I thought the filters panel with tags and categories worked quite well. We should have a call to discuss things, I'll attempt an early morning.

1.2 The departments should be a single selection not a multi selection.

1.3 Are you not able to make dynamically:

1.3.1 As many root level dropdowns that are in the JSON (Categories, Colleges and Tags)? An empty one as the first entry in the dropdown for nothing selected.

1.3.2 If an entry is selected in Colleges then dynamically, move everything down and show a Departments label and dropdown? It would be valid to just select a college without a department.

1.3.3 If Athletics is selected do the same as 1.3.2 for Athletics, adding a new label and dropdown?

1.3.4 If this is a No to 1.3 then fine, c'est la vie, and any ideas Misho?

**Comment by mihail-varbanov** [comment]
**Re: University Monday 2023/01/30** / 2022-01-31

1.1 I really like what you did with the filters dynamically, now it works just as I would expect. Thank you.

1.2 Could you add a new "root level" filter of "RSO" with values of "Yes" and "No" please.

1.3 Item 1.2 makes me wonder if we can support checkboxes in addition to dropdowns, something for you to think on, where we would just have a label "RSO" with a checkbox instead of a dropdown.

1.4 We can discuss this in more detail but the files in s3 for categories and tags should be retired and we should all start using filters which will be much more powerful. Sandeep does not know this yet, but over the next year, maybe less, we will be replacing his Events BB with a better version in GO by Petyo, so the filters will then be under our control.

**Comment by mihail-varbanov** [comment]
**Re: University Monday 2023/01/23** / 2023-01-27

> 1.3 We like the idea of adding the Categories and Tags I mentioned into filters also. Where do we get the content for these please, API calls for each etc?

MV: I would not hurry to rush into this so quickly. Things only look good on the surface and if you dive into details you will see that this is not so obvious and easy.

We get these both from Events service. Please take a look at what we get from the backend:
https://rokwire-ios-beta.s3.us-east-2.amazonaws.com/Assets/event.categories.json
https://rokwire-ios-beta.s3.us-east-2.amazonaws.com/Assets/event.tags.json

**Categories**
Categories apply very well to become a filter but I see some particularities that should be resolved before diving into coding:

Athletics subcategories
This is not exactly a problem but rather a warning light. Fortunately we do not use Athletics subcategories anywhere in the app, the client loads sports from sports service. Subcategories like these should be represented as nested/dependent filter types, like the department depends on college in your example. But the difference is that they should be visible only if the Athletics category is selected because it would not be quite user-friendly to keep this control permanently visible and mostly disabled when other categories are selected. The behavior of such a subfilter differs from what we currently have - the filter is permanently visible and just updates its inner/dropdown content when the parent filter changes its selection.

Usage outside home panel filters
We display categories in group card heading, in group details panel and anywhere else. If we want to keep our current UI we should somehow mark the category filter indicating that its should be displayed where we currently display the group's category property. Or, we can display all group filter values listed consequently (e.g. Category, College, Department if these are selected) but often there is not much room for this, especially in the card widget. BTW, I rather tend to attempt to list them all.

Backward compatibility
We have a number of existing groups that use the standalone "category" field and have no filters set. If the client simply stops to acknowledge the existing standalone group category and works with filters/attributes only, we will lose all current group categories. Best of all would be if we duplicate these categories as filter entries, one-time on the backend, forget the current standalone category field and work only with filters. However if this is not available it would be a small nightmare in the client app to support both category sources/formats.

**Tags**
In my opinion tags are something different than what we represent in filters/attributes.

Size
We currently have 337 tag definitions. Obviously it is not a good idea to fill them as dropdown content and to expect the user to scroll through them. We should think of some UI that is similar to our "Group Tags" selection page that has advanced lookup capabilities. If we are going to represent tags as filters/attributes item we should design a way to mark such a filter that it needs some different/special handling than a regular filter with simple dropdown. This will increase the complexity of filters/attributes and I am sure that it will bring more problems than an actual benefit.

Logic
I do not know for you but I think that our tags logic is mistaken. Normally users can define their own, custom entered arbitrary tags, and the list of tags is just a helper to identify the existing tags. When a new tag is defined for the first time then it should appear in the helpers list. This is something very different from filters/attributes logic (but I consider that our current tags logic can be accomplished).

Backward compatibility
Basically the same concerns like categories.

So my conclusion is that we would better leave tags as they are, or even improve them to the typical tags logic. Also, if we decide to list all filter values in groups UI (cards, detail panels) like we do with categories now, tags will be an exception and needs to be handled separately. For example, it is good to list "Academics" / "Applied Health Sciences" / "Fine & Applied Arts" in card/detail heading but as regards to tags they either should not be listed at all, or should be displayed separately and differently formatted, e.g "Tags: art and music, open house, performance, webinar"

**Comment by mihail-varbanov** [comment]
**Re: University Monday 2023/01/30** / 2023-02-01

1.1 We should put attributes/filters into the Modify/Create of Research Groups also.

1.2 Answers to Friday's comments:

1.2.1 "Usage outside home panel filters". I agree with "...BTW, I rather tend to attempt to list them all." If they exist, can we omit so we do not take up extra lines?

1.2.2 I would show Category in the same place as we do now and I would place {College}\r\n{Department} above the "Updated about ..." time stamp.

1.2.3 I would place "RSO" in Orange like we do for "Private" or "Admin" etc to the right of the Category on the same line. Not right justified, just after the Category: Other RSO

1.2.4 I agree on the backend processing all existing Categories and Tags so we can remove the old way of doing it.
